repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -0,0 +1,129 @@
+                                     QueryBuilder queryBuilder) {
+        this.rowMapper = rowMapper;
+        this.namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
+        this.queryBuilder = queryBuilder;
+    }
+
+    @Override
+    public List<Certificate> find(SearchCriteria searchCriteria) {
+        MapSqlParameterSource parameters = new MapSqlParameterSource();
+        List<String> whereClauseFindByParameters = new ArrayList<>();
+        List<String> orderByAscDesc = new ArrayList<>();
+
+        queryBuilder.nameOfATag(whereClauseFindByParameters, parameters, searchCriteria.getTagName());
+        queryBuilder.nameOfCertificate(whereClauseFindByParameters, parameters, searchCriteria.getCertificateName());
+        queryBuilder.descriptionOfCertificate(whereClauseFindByParameters, parameters, searchCriteria.getCertificateDescription());
+        queryBuilder.sortByCertificateName(orderByAscDesc, searchCriteria.getSortByName());
+        queryBuilder.sortByCertificateCreationDate(orderByAscDesc, searchCriteria.getSortByCreateDate());
+
+        String finishedQuery = queryBuilder.sqlBuild(whereClauseFindByParameters, orderByAscDesc, SELECT_ALL_CERTIFICATES);
+
+        return namedParameterJdbcTemplate.query(finishedQuery, parameters, rowMapper);
+    }
+
+    @Override
+    public Optional<Certificate> findById(Long id) {
+        SqlParameterSource parameterSource = new MapSqlParameterSource().addValue(ID_PARAMETER, id);
+        List<Certificate> certificates = namedParameterJdbcTemplate.query(SELECT_CERTIFICATE_BY_ID, parameterSource, rowMapper);
+
+        return certificates.stream().findFirst();
+    }
+
+    @Override
+    public boolean attachTag(long certificateId, long tagId) {
+        int update;
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue(GIFT_CERTIFICATE_ID_PARAMETER, certificateId)
+                .addValue(TAG_ID_PARAMETER, tagId);
+
+        try {
+            update = namedParameterJdbcTemplate.update(INSERT_TAG_TO_CERTIFICATE, parameterSource);
+        } catch (DataAccessException e) {
+            throw new DataException("There is a problem to attach: Tag id - " +
+                    tagId + ", Certificate id - " + certificateId, e);
+        }
+        return update > 0;
+    }
+
+    @Override
+    public boolean detachTag(long certificateId, long tagId) {
+        int update;
+        SqlParameterSource parameterSource = new MapSqlParameterSource()
+                .addValue(GIFT_CERTIFICATE_ID_PARAMETER, certificateId)
+                .addValue(TAG_ID_PARAMETER, tagId);
+
+        try {
+            update = namedParameterJdbcTemplate.update(DELETE_TAG_FROM_CERTIFICATE, parameterSource);
+        } catch (DataAccessException e) {
+            throw new DataException("There is a problem to delete: Tag id - " +
+                    tagId + ", Certificate id - " + certificateId, e);

```

Comment:
According to the requirements, you need localization in your application

Reply:
fixed